### ISSUE TITLE
Undo and redo operations

### ISSUE THREAD

**Issue by aviad-ariel** [wbkd/react-flow#656]
Is there any support of undo and redo operations? 
couldn't find anything related.

Great library by the way!

**Comment by chrtze** [comment]
hi, there is no built-in support at the moment, but it's a good idea. any suggestions on how an api for this could look like? maybe it is already possible by using the internal state: https://reactflow.dev/docs/api/internal-state-actions/

**Comment by aviad-ariel** [comment]
I will try it on the weekend and let you know, if anyone has any ideas/suggestions that would be great.

**Comment by obaid** [comment]
Interested in this as well. We can probably use https://www.npmjs.com/package/undo-peasy to implement it.

**Comment by harshambastha** [comment]
Is it implemented?

**Comment by aronm99** [comment]
I am interested in this feature as well! I am wondering if there has been any progress on this feature or if not any more ideas on how to implement it.

**Comment by pnkflydgr** [comment]
The SaveRestore component in the examples shows you how you can implement this from an external store perspective.

Would be nice if it was part of the module tho.

**Comment by pete1019** [comment]
Anyone got undo / redo working and can help, please? Thanks a lot.

**Comment by xplato** [comment]
@pete1019 My team is working on a package exporting a React Hook to handle this exact use-case, tailored for usage with ReactFlow (although not limited to it). I'll provide an update here once we've finished it.

All of the undo/redo packages we've come across have been insufficient for our use-case in ReactFlow. In particular, with multiple functional state changes. That is, similar to `setElements(els => do.something(els));`

**Comment by pete1019** [comment]
> 

@xplato thx for the reply.
When is the Estimated Time To Complete?

**Comment by xplato** [comment]
@pete1019 It's live! You should be able to use it without any issues. If you encounter any, however, please email us (email found at https://github.com/Infinium8).

[use-undoable on NPM](https://www.npmjs.com/package/use-undoable)

**Comment by pete1019** [comment]
@xplato THANKS!
Would you be able to share a video where this can be seen in action with reac-flow?

**Comment by xplato** [comment]
@pete1019 Yes, absolutely. I will do so tomorrow. It's quite late here. Thanks!

**Comment by xplato** [comment]
Better than a video, here's a live example using the new `useUndoable` hook: https://codesandbox.io/s/react-flow-undo-7bpqd

Move a few nodes around and click the undo/redo buttons.

**Comment by pete1019** [comment]
@xplato cool thx.
How about the connections between the nodes?

And what if i get information from Database and show/change them as information (Text) inside a node? Would your undo/redo be capable of providing undo for that text as well?

**Comment by xplato** [comment]
Think of the `useUndoable` hook like `useState`. It simply keeps track of all changes made to the state and exports helper functions (`undo`, `redo`, etc.) which allow you to modify the history of those changes.

Therefore, this hook is capable of managing anything you decide to put into it. All you have to do is setup event handlers exported by ReactFlow and then modify your local state accordingly. Here's a minimal example:

```
const MyFlow = () => {
	const [
		elements,
		setElements,
		{
			undo,
			redo
		}
	] = useUndoable([]);

	return (
		<ReactFlow
			elements={elements}
			onNodeDragStop={(ev, node) => setElements(
				// modify your elements here
			)}
			onConnect={(ev, node) => setElements(
				// modify your elements here
			)}
			onEdgeUpdate={(ev, node) => setElements(
				// modify your elements here
			)}
		/>
	);
}
```

**Comment by carlosjedwab** [comment]
@xplato awsome feature to have in the flow!

However, while using it, it seems that every change to the elements is delayed...

Even in that codesandbox you sent (https://codesandbox.io/s/react-flow-undo-7bpqd), just by putting a console.log(elements) inside the onNodeDragStop, it shows the elements of a previous change. Look at the printed positions for example, they won't match with the nodes position ate the screen. Is there any solutions for this?

```
onNodeDragStop={(ev, node) =>
          setElements((els) => {
            console.log(els);
            return els.map((e) => {
              if (e.id === node.id) {
                let n = {
                  ...elements.filter((e) => e.id === node.id)[0],
                  position: node.position
                };

                return n;
              }

              return e;
            });
          })
        }
```

**Comment by carlosjedwab** [comment]
@xplato awsome feauture!

However while using it, it seems something went wrong with my layouting. When using the `useUndoable` instead of the `useState`, the layouting of the nodes is not rendered right away (only after some other change to the elements). I suspect the layouting changes are not notified correctly.

I solved this by adding and removing a mock node right away after the layouting is done, but I don't beleve this is a very good hack. Is there a better way to solve this issue, and is this connected to the `Math.random() / 1000` I have to add to the position while getting the layouted elements?

**Comment by xplato** [comment]
@carlosjedwab Yes, I'm noticing the same issues.

I'm trying to locate the issue within the code. Since ReactFlow depends on deep state updates (not direct mutation), I believe it may be a problem with how updates are sent to the hook.

I've [opened an issue](https://github.com/Infinium8/useUndoable/issues/1) within my lib; you can track the progress there.

**Comment by xplato** [comment]
@carlosjedwab I've spent a few hours testing the `useUndoable` hook with ReactFlow and I'm unable to replicate either your issue or the one I observed. Could you create a Codesandbox and link to it in the [issue I created here?](https://github.com/Infinium8/useUndoable/issues/1) Thank you.

**Comment by carlosjedwab** [comment]
@xplato I posted it there in the issue, but here it is in this discussion as well: https://codesandbox.io/s/react-flow-undo-forked-wn2kv?file=/src/App.js

In App.js, there are two sections for declaring the elements:

```
// Uncomment this section to allow the undo/redo feature
  const [elements, setElements, { undo, canUndo, redo, canRedo }] = useUndoable(
    initialElements
  );

  // Uncomment this section to not allow the undo/redo feature
  //const [elements, setElements] = React.useState(initialElements);
  //const undo = () => {};
  //const canUndo = false;
  //const redo = () => {};
  //const canRedo = false;
```

If the **first one is uncommented and the second one is commented,** the undo/redu feature will work, but the layouting error will apear. 
If the **first one is commented and the second is uncommented**, then the undo/redo feature is not activated, but the layouting works just fine.

_(in the second one, the element is a state, and the undo/redo functions are mocked)_

**Comment by xplato** [comment]
@carlosjedwab Thank you! [Version 3.2.1](https://github.com/Infinium8/useUndoable/commit/cb22f5b32584d583c7c53f851af8f8a33bf8cb20) fixes this issue.

**Comment by carlosjedwab** [comment]
@xplato Hey, so, the bug was fixed! But sorry to bother you again, there seems to be a secondary bug:

In the same Codesandbox https://codesandbox.io/s/react-flow-undo-forked-wn2kv?file=/src/App.js, if you drag the nodes around and then undo, the elements are in fact undone.

However, if you layout the elements **and then** undo, the layouting is not undone. I would guess the issue is that the layouting is not detetected as a change to be undone/redone.

**Comment by xplato** [comment]
> But sorry to bother you again

Haha, not at all. You're being very helpful; showing me bugs and the like.

> However, if you layout the elements and then undo, the layouting is not undone. I would guess the issue is that the layouting is not detetected as a change to be undone/redone.

Thanks for letting me know. I should have tested more thoroughly. I'll get right on this.

**Comment by carlosjedwab** [comment]
@xplato Just found another major bug when you use the useUndoable. I looks like a function that uses setElements internally in a custom node is not working correctly.

For example, if you have a custom node with an input value `value` and a function `onChange` that, when defined, uses `setElements` to update the `value`. when this onChange is called, it doesn't consider any elemnt create after the crating of the given custom node.

Was it possible to understand the problem? If not, I can create a codesandbox

**Comment by pete1019** [comment]
> 

@carlosjedwab would you mind opening the major bug here as well?
https://github.com/Infinium8/useUndoable/issues
thx

**Comment by pete1019** [comment]
@carlosjedwab 
are you using useUndoable in production now? Everything working fine?

**Comment by xplato** [comment]
@pete1019 Just a heads up, the previous issue explained here wasn't caused by useUndoable. It was a conflict with Dagre in some form or another. 

So far, I haven't had any major issues.

**Comment by carlosjedwab** [comment]
@pete1019 
I'm currently not using useUndoable, the project ended up in a halt

**Comment by pete1019** [comment]
@carlosjedwab you mean your project is in a halt?

**Comment by xplato** [comment]
@pete1019 @carlosjedwab 

If you guys encounter any issues with my package, I'd look into it right away. I haven't been pushing because, so far, it does everything (and more) than it needs to, so there's no reason to make changes that don't add any value.

I haven't heard of any issues with it besides the ones explained here; which were either resolved within a day or two or were not caused by my package. In fact, I'm currently using it in production without any issues.

That's all. I just wanted to clarify that I'll be on top of any issues, there just haven't been any.

**Comment by carlosjedwab** [comment]
@pete1019 Going into some context, the project was a school project, and the semester already ended hahaha. It was a success

**Comment by raarts** [comment]
@chrtze Could you please mention the `use-undoable` package created by @xplato in the README? I spent  too much time implementing undo/redo, because I didn't know that package existed. Thanks!

**Comment by xplato** [comment]
@raarts There is a React Flow example directly in my package and/or via the CodeSandbox links in this thread. Adding my third-party package to React Flow is completely unnecessary.

**Comment by pete1019** [comment]
> @raarts There is a React Flow example directly in my package (see the [`react-flow-example`](https://github.com/Infinium8/useUndoable/tree/main/react-flow-example) folder) and/or via the CodeSandbox links in this thread. Adding my third-party package to React Flow is completely unnecessary.

i think all he is asking is:
"use-undoable" should be mentioned under "Community Packages" in README.md
Which would be great.

**Comment by moklick** [comment]
Good point. We will add that!

**Comment by xplato** [comment]
@pete1019 @raarts Oh, I see. I apologize; I didn't know this area existed in the README. Thanks for pointing this out!

**Comment by moklick** [comment]
added to the Readme. Thanks @xplato for this great work :)

**Comment by arthurodriguesbatista** [comment]
does anyone know how I can avoid moving a node counting as a series of state updates. Each instance of dragging is updating the history state as mentioned here: https://github.com/Infinium8/useUndoable#integration-with-react-flow-v10

**Comment by DraksBG** [comment]
@xplato hey there do you have any showcase for the new react flow-v10 update how the "use-undoable"  should be working. My team is trying to use it but we have found our self in quite a tricky situation with the re render of the main component. I would appreciate a small demo . Thanks in advance !

**Comment by xplato** [comment]
@DraksBG Sure!

So there are two very common issues I've seen:

1. You have a mismatch of state updates because you are using more than one state manager. This is very common, especially when you try to have two separate `useUndoable` (or `useState`) calls, one for nodes and one for edges. This can very quickly become hard to manage and will introduce a lot of small bugs that are difficult to see and test. The only solution is to use just one state and split it up internally ([example](https://github.com/Infinium8/useUndoable/issues/11#issuecomment-1204123860))
2. The new "generic" functions in v10 (`onNodesChange` and `onEdgesChange`) are propagating events that you don't need, leading to a series of illogical state mutations. The most common example of this is moving nodes: it does not count as one update, but a _series_ of updates, one for each pixel you move (more or less). If you suffer from this, [here](https://codesandbox.io/s/react-flow-v10-undo-sjy7tu?file=/src/App.js) is a good way to get around this, using the generic functions. Another solution is to use the more specific events and handle those explicitly. For example, using `onConnect` and `onNodeDragStop` instead of the generic `onNodesChange` function.

I'm not sure what you mean by "tricky situation with the re render of the main component" exactly, but I hope it fits into one of these categories, as there are solutions.

Let me know if I can provide more information here. Best of luck to you!

**Comment by moklick** [comment]
thanks for your help here @xplato 👍

**Comment by DraksBG** [comment]
@xplato Thank our team will try especially with the trigger event, I have a quick question on that if it is put inside the onNodesChanges could it create a side effect when we use trigger event as a dependency ?
Which is better as per your view  onNodeDragStop or onNodeChange ?

Thank you again !

**Comment by raarts** [comment]
My $0.02: `onNodeDragStop` is better. `onNodeChange` is just too noisy.
I use the `use-undo` package which I found to be a little more intuitive to use.

**Comment by carboni-rob** [comment]
When deleting nodes, the `onEdgesChange` is also called with `remove` changes to the orphaned edges.
This is a problem when using `undoable` - and in general I believe in a controlled flow.
If we have
```ts
  const [{ nodes, edges }, setState, { undo, redo }] =
    useUndoable<EditorState>({
      nodes: initialNodes,
      edges: initialEdges
    });

  const onNodesChange =
    (changes: NodeChange[]) => {
      setState(
        (currentState) => {
          return {
            ...currentState,
            nodes: applyNodeChanges(changes, currentState.nodes),
          };
        });
    };

  const onEdgesChange =
    (changes: EdgeChange[]) => {
      setState((currentState) => {
        return {
          ...currentState,
          edges: applyEdgeChanges(changes, currentState.edges),
        };
      });
    };
```
Looks like when deleting a node the following chain of events happens:
- `onNodesChange` fires, detects orphaned nodes, calls `onEdgesChange` and waits for its execution (note at this point `currentState` in `onNodesChange` still has the orphaned edges);
- `onEdgesChange` executes and correctly updates the edges state;
- `onNodesChange` completes execution, overwriting the state with the now stale edges state (result: orphaned edges are added back in).
Would it be better NOT to check for orphaned nodes in a controlled flow and let the developers manually check for orphaned nodes?
The workaround I found consists in creating two additional pieces of state for nodes and edges, update them individually, call `setState` every time they change in an effect hook.

**Comment by atruttmann** [comment]
I'm noticing the exact same issue as @carboni-rob. Orphaned edges are added back into my state after a node is deleted. Tagging @xplato for visibility :)

**Comment by xplato** [comment]
I can take a look at this in more detail later @atruttmann, @carboni-rob--but in the meantime, have you tried playing with the `ignoreAction` / `ignoreIdenticalMutations` options?

**Comment by pete1019** [comment]
> 

@raarts would you please share more information or code how you implemented use-undo to reactFlow? Thanks.

**Comment by raarts** [comment]
This is basically how I did it:
```TypeScript

const MyComponent = () => {
  const [keepState, setKeepState] = useState(0);
  const [nodes, setNodes] = useState<Node[]>([]);
  const [edges, setEdges] = useState<Edge[]>([]);
  const [{ present }, { set, reset, undo, redo, canUndo, canRedo }] = useUndo<UndoData>({
    nodes: [],
    edges: [],
  });
  useEffect(() => {
    if (keepState === 0) {
      reset({ nodes, edges });
    } else {
      set({ nodes, edges });
    }
  }, [keepState]);

  return (
    <GraphEditor
      {...props}
      nodes={nodes}
      setNodes={setNodes}
      edges={edges}
      setEdges={setEdges}
      keep={() => setKeepState((keepState) => keepState + 1)}
      undo={undo}
      redo={redo}
      canUndo={canUndo}
      canRedo={canRedo}
    />
  )
}
```

This would give me a keep() function I could call the save the current state for undoing later.
I bound the undo/redo to buttons.

**Comment by dominiczy** [comment]
Has anyone tried using [zundo](https://github.com/charkour/zundo) for this? I'd like to add undo functionality and our app already uses `zustand` with a controlled flow

**Comment by prabak** [comment]
@dominiczy were you able to get zundo to work with controlled reactlfow?

**Comment by dominiczy** [comment]
@prabak yeah it works reasonably well for me. Just need to take care to pause it during Dragging operations, else you'll have a lot of intermediate states

**Comment by Ashwin746544** [comment]
@dominiczy can you please share your code with zundo + zustand controlled flow , I need it urgently so , Thanks in advance

**Comment by kyh0703** [comment]
> @prabak yeah it works reasonably well for me. Just need to take care to pause it during Dragging operations, else you'll have a lot of intermediate states

Thank you for your kind word!

**Comment by DatTN95** [comment]
> zundo

hi you, can you guide me?

**Comment by youngjuning** [comment]
How to solve add Nodes event
